Documentation/proposals/202310-automated-sharding.md

```diff
@@ -0,0 +1,164 @@
+
+To address these issues, this document outlines a solution to enhance the Prometheus Operator by enabling the usage of Horizontal Pod Autoscalers(HPAs) to scale shards up and down. This improvement aims to simplify the process of scaling Prometheus instances, making it more accessible and efficient for operators.
+
+# Pitfalls of the current solution
+
+Prometheus can scale vertically pretty well, but instances that grow up to hundreds of GiB/TiB in WAL size start to get troublesome:
+
+* Eventual cardinality explosions have a big blast radius if a single Prometheus instance is responsible for scraping the majority of monitored applications.
+* Recovering from a crash takes several minutes due to the WAL replay.
+
+Meanwhile, another strategy would be to use Horizontal Pod Autoscalers (HPAs) with Prometheus statefulsets by increasing the number of replicas. However, arguably it makes things more complicated:
+
+* Replicas with the same scrape configuration will not shard the scraping load into different instances. Instead, we get several Prometheus instances generating duplicated metrics.
+
+# Goals
+
+* Enable scaling of the Prometheus shards up and down via Horizontal Pod Autoscaler objects.
+
+# Non-Goals
+
+* Implement the autoscaling logic into the Prometheus operator.
+* Automatic balancing of targets to ensure homogeneous load across all Prometheus pods.
+
+# Audience
+
+* Users looking for a shard autoscaling mechanism with a good cost-reliability ratio.
+
+# How
+
+Today, there are a few strategies to measure load on top of Prometheus instances.
+
+1. Perhaps you can use [Kubernetes Metrics API](https://kubernetes.io/docs/tasks/debug/debug-cluster/resource-metrics-pipeline/#metrics-api) to measure load in terms of CPU and/or Memory usage.
+
+2. Another strategy might be to measure the amount of samples ingested per second
+
+- Prometheus Agents: `rate(prometheus_agent_samples_appended_total[5m])`
+- Prometheus Servers: `rate(prometheus_tsdb_head_samples_appended_total[5m])`
+
+3. Or even measuring samples queried per second for Prometheus instances under heavy query load.
+
+- Prometheus Servers: `rate(prometheus_engine_query_samples_total[5m])`
+
+Modern Custom Horizontal Pod Autoscalers (HPAs), such as Keda are already capable of [scaling CRDs](https://keda.sh/docs/2.11/concepts/scaling-deployments/#scaling-of-custom-resources) based on different parameters, such as [CPU, Memory](https://keda.sh/docs/2.11/scalers/cpu/) or even [custom Prometheus metrics](https://keda.sh/docs/2.11/scalers/prometheus/).
```

Comment:
The main point here IMHO is to say that the operator is going to be "autoscaler-agnostic". We don't prescribe that you use HPA or Keda: as long as the scaler supports the scale subresource, it should work.

Reply:
Good point, I can see that the current wording is too strong towards Keda. I'll reword it